Shell/main.cpp

```diff
@@ -300,16 +300,60 @@ static Vector<String> expand_globs(const StringView& path, const StringView& bas
     return res;
 }
 
+static Vector<String> expand_parameters(const StringView& param)
+{
+    Vector<String> res;
+
+    if (param.length() > 1 && param[0] == '$') {
+        String param_name = String(param.substring_view(1, param.length() - 1));
+
+        if (param_name == "?") { //Exit status of most recently executed
+            res.append(String::format("%d", g.last_return_code));
+        } else {
+            char* value = getenv(param_name.characters());
+
+            if (value == nullptr) {
+                res.append("");
+            } else {
+                String str_value = String(value);
+                int last_char = 0;
+
+                // This ignores extra spaces between text, because that's what bash seems to do
+                for (int i = 0; i < str_value.length(); ++i) {
+                    if (str_value.characters()[i] == ' ') {
+                        res.append(str_value.substring(last_char, i - last_char));
+                        last_char = i + 1;
+                    } else if (i == (str_value.length() - 1)) {
+                        // Handles when we get to the end of the string.
+                        res.append(str_value.substring(last_char, i - last_char + 1));
+                    }
+                }
+            }
+        }
+    } else {
+        res.append(param);
```

Comment:
Style nitpick: it's rather hard to see what `if` this `} else {` refers to. Instead, it'd be better to avoid deep nesting and use early returns for the "uninteresting" paths. It could look something like this:

```c++
bool is_variable = param.length() > 1 && param[0] == '$';
if (!is_variable)
    return { param };

String param_name = ...;
if (param_name == "?")
    return { ... };

char* value = getenv(param_name.characters());
if (value == nullptr)
    return { "" };

Vector<String> res;
for (auto& part : String(value).split_view(' '))
    res.append(part);
return res;
```